lib/include/currentSource.h

```diff
@@ -0,0 +1,110 @@
+    {
+        setVarMode(varName, enabled ? VarMode::LOC_ZERO_COPY_INIT_HOST : VarMode::LOC_HOST_DEVICE_INIT_HOST);
+    }
+
+    //! Set variable mode of neuron model state variable
+    /*! This is ignored for CPU simulations */
+    void setVarMode(const std::string &varName, VarMode mode);
+
+    void initDerivedParams(double dt);
+
+    //------------------------------------------------------------------------
+    // Public const methods
+    //------------------------------------------------------------------------
+    const std::string &getName() const{ return m_Name; }
+
+    //! Gets the current source model used by this group
+    const CurrentSourceModels::Base *getCurrentSourceModel() const{ return m_CurrentSourceModel; }
+
+    const std::vector<double> &getParams() const{ return m_Params; }
+    const std::vector<double> &getDerivedParams() const{ return m_DerivedParams; }
+    const std::vector<NewModels::VarInit> &getVarInitialisers() const{ return m_VarInitialisers; }
+
+    int getClusterHostID() const{ return m_HostID; }
+
+    int getClusterDeviceID() const{ return m_DeviceID; }
+
+    bool isZeroCopyEnabled() const;
+    bool isVarZeroCopyEnabled(const std::string &var) const{ return (getVarMode(var) & VarLocation::ZERO_COPY); }
+
+    //! Get variable mode used by current source model state variable
+    VarMode getVarMode(const std::string &varName) const;
+
+    //! Get variable mode used by current source model state variable
+    VarMode getVarMode(size_t index) const{ return m_VarMode[index]; }
+
+    void addExtraGlobalParams(std::map<std::string, std::string> &kernelParameters) const;
+
+    //! Does this current source require any initialisation code to be run
+    bool isInitCodeRequired() const;
+
+    //! Does this current source require an RNG to simulate
+    bool isSimRNGRequired() const;
+
+    //! Does this current source group require an RNG for it's init code
+    bool isInitRNGRequired(VarInit varInitMode) const;
+
+    //! Is device var init code required for any variables in this current source
+    bool isDeviceVarInitRequired() const;
+
+    //! Can this current source run on the CPU?
+    /*! If we are running in CPU_ONLY mode this is always true,
+        but some GPU functionality will prevent models being run on both CPU and GPU. */
+    bool canRunOnCPU() const;
+
+private:
+    //------------------------------------------------------------------------
+    // Members
+    //------------------------------------------------------------------------
+    std::string m_Name;
+
+    const CurrentSourceModels::Base *m_CurrentSourceModel;
+    std::vector<double> m_Params;
+    std::vector<double> m_DerivedParams;
+    std::vector<NewModels::VarInit> m_VarInitialisers;
+
+    //!< Whether indidividual state variables of a neuron group should use zero-copied memory
+    std::vector<VarMode> m_VarMode;
+
+    //!< The ID of the cluster node which the neuron groups are computed on
+    int m_HostID;
```

Comment:
This MPI stuff is a bit half-finished at best (it was the result of a previous GSoC project) but I'm pretty sure current sources shouldn't have a host and device ID - they should **always** be simulated on the same device and host the neuron they're being injecting into is located on